File: src/logic/reference/matching.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <algorithm>
#include <iterator>

/**
 * Coverage Score in C++: |A ∩ B| / |B|
 * Measures what fraction of the job's REQUIRED skills the student covers.
 * Unlike Jaccard, does NOT penalise knowing extra skills.
 */
double calculateCoverage(std::set<std::string> studentSkills, std::set<std::string> requiredSkills) {
    if (requiredSkills.empty()) return 0.0;

    std::vector<std::string> intersect;
    std::set_intersection(studentSkills.begin(), studentSkills.end(),
                          requiredSkills.begin(), requiredSkills.end(),
                          std::back_inserter(intersect));

    return (double)intersect.size() / (double)requiredSkills.size();
}

/**
 * Binary Search on sorted vector of objects
 */
struct Internship {
    std::string company;
    std::string role;
};

int binarySearch(const std::vector<Internship>& internships, std::string target) {
    int low = 0;
    int high = internships.size() - 1;

    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (internships[mid].company == target) return mid;
        if (internships[mid].company < target) low = mid + 1;
        else high = mid - 1;
    }
    return -1;
}

int main() {
    std::set<std::string> student = {"React", "JavaScript"};
    std::set<std::string> job = {"React", "JavaScript", "Tailwind"};
    
    // student covers 2 of 3 required skills => 0.666...
    std::cout << "Coverage Score: " << calculateCoverage(student, job) << std::endl;
    return 0;
}

File: src/logic/reference/matching.py
import math

def coverage_score(student_skills: set, required_skills: set) -> float:
    """
    Coverage Score: |A ∩ B| / |B|
    Measures what fraction of the job's REQUIRED skills (B) the student (A) covers.
    Unlike Jaccard, does NOT penalise students for knowing extra skills.
    """
    if not required_skills:
        return 0.0
    intersection = len(student_skills.intersection(required_skills))
    return intersection / len(required_skills)

def binary_search_internships(sorted_internships, target_company):
    """
    Binary search for an internship by company name
    Complexity: O(log n)
    """
    low = 0
    high = len(sorted_internships) - 1
    
    while low <= high:
        mid = (low + high) // 2
        mid_company = sorted_internships[mid]['company'].lower()
        
        if mid_company == target_company.lower():
            return sorted_internships[mid]
        elif mid_company < target_company.lower():
            low = mid + 1
        else:
            high = mid - 1
    return None

# Example usage
student_skills = {"Python", "SQL", "Data Analysis"}
internship_skills = {"Python", "PyTorch", "SQL", "Machine Learning"}

# Coverage: student covers 2/4 required skills = 0.50
score = coverage_score(student_skills, internship_skills)
print(f"Coverage Score: {score:.2f}")
